java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+  TempDirHijackingFullPath() { this = "TempDirHijackingFullPath" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    isSourceUnified(source, state)
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    isAdditionalTaintStepUnified(node1, state1, node2, state2)
+  }
+
+  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+    // From a `File::delete` call, to a call to `File::mkdir(s)`
+    isNonThrowingDirectoryCreationExpression(sink.asExpr(), _) and
+    state instanceof FromDeleteFileFlowState
+  }
+}
+
+/**
+ * A configuration that tracks the full path of the temporary directory taint all the way to an
+ * `unsafeUse` of the potentially hijacked temporary directory.
+ *
+ * This is intentionally not used in the the generation of the displayed path;
+ * this is because there may not exist an explicit path from the call to `File::mkdir(s)` call to the the `unsafeUse`.
+ */
+class TempDirHijackingFullPathInlcudingUnsafeUse extends TaintTracking2::Configuration {
+  TempDirHijackingFullPathInlcudingUnsafeUse() {
+    this = "TempDirHijackingFullPathInlcudingUnsafeUse"
+  }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    isSourceUnified(source, state)
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    isAdditionalTaintStepUnified(node1, state1, node2, state2)
+    or // `File::mkdir(s)` is not an end-state when looking for an unsafe use
+    isNonThrowingDirectoryCreationExpression(node1.asExpr(), _) and
+    node1.asExpr() = node2.asExpr() and
+    state1 instanceof FromDeleteFileFlowState and
+    state2 instanceof FromMkdirsFileFlowState
+  }
+
+  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+    // From a `File::delete` call, to a call to `File::mkdir(s)`
+    isUnsafeUseUnconstrainedByIfCheck(_, sink.asExpr()) and
+    state instanceof FromMkdirsFileFlowState
+  }
+}
+
+from
+  DataFlow::PathNode pathSource, DataFlow::PathNode pathSink, DataFlow::Node deleteCheckpoint,
+  MethodAccess creationCall, Expr unsafe
+where
+  // Generate the full path to display to the user
+  any(TempDirHijackingFullPath c).hasFlowPath(pathSource, pathSink) and
+  // Find the delete checkpoint to display to the user
+  any(TempDirHijackingToDeleteConfig c).hasFlow(pathSource.getNode(), deleteCheckpoint) and
+  // Find the delete checkpoint to display to the user
+  any(TempDirHijackingFromDeleteConfig c).hasFlow(deleteCheckpoint, pathSink.getNode()) and
+  // Find a full path where an `unsafe` use is present
+  any(TempDirHijackingFullPathInlcudingUnsafeUse c)
+      .hasFlow(pathSource.getNode(), DataFlow::exprNode(unsafe)) and
```

Comment:
Is it fine that I'm using this many taint tracking and data flow modules in one query? I'm concerned about the performance impact of this potentially.

Reply:
Nothing inherently terrible about it. We'll do perf analysis prior to committing anything to check.